subiquity/server/controllers/zdev.py

```diff
@@ -607,20 +643,60 @@ def __init__(self, app):
                 zdevinfos = [ZdevInfo.from_row(row) for row in devices]
             self.zdevinfos = OrderedDict([(i.id, i) for i in zdevinfos])
 
+    def load_autoinstall_data(self, data: ZdevAi) -> None:
+        self.ai_data = data
+
+    @with_context()
+    async def apply_autoinstall_config(self, context) -> None:
+        if self.zdev_handling_task is not None:
+            await self.zdev_handling_task
+
+    def start(self) -> None:
+        if self.ai_data:
+            self.zdev_handling_task = schedule_task(self.handle_zdevs())
+
+    def make_autoinstall(self) -> ZdevAi:
+        # Small "optimization" to avoid producing a config that enables or
+        # disables a given device multiple times in a row.
+        return [x for x, _ in itertools.groupby(self.done_actions)]
```

Comment:
praise: TIL about `itertools.groupby` (very nifty) and this is much better than our original plan of putting some of the logic in `handle_zdevs`. Thanks!

Reply:
Agreed, it makes the code cleaner in the `chzdev` function, which is more important :)